sources/api/netdog/systemd-derive/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+            .expect("Will never be anything but a struct")
+            .fields
+            .iter()
+            .filter_map(|f| f.ident.clone())
+            .collect();
+
+        tokens.extend(quote! {
+            impl std::fmt::Display for #ident {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    // `Vec`s and `Option`s are both iterators, allowing us to write the same code
+                    // for both
+                    #(for section in self.#sections.iter() {
+
+                            write!(f, "{}", section.to_string())?;
+                    })*
+                    Ok(())
+                }
+            }
+        });
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// A macro to simplify serializing a section of a unit file.  See the description in the lib
+/// documentation or the README.
+#[proc_macro_derive(SystemdUnitSection, attributes(systemd))]
+pub fn derive_systemd_unit_section(input: TokenStream) -> TokenStream {
+    // Parse the AST and "deserialize" into SystemdUnitSection
+    let ast = parse_macro_input!(input as DeriveInput);
+    let n = SystemdUnitSection::from_derive_input(&ast)
+        .expect("Unable to parse `systemd` macro arguments");
+
+    quote!(#n).into()
+}
+
+#[derive(Debug, FromDeriveInput)]
+#[darling(supports(struct_named))]
+#[darling(attributes(systemd))]
+struct SystemdUnitSection {
+    pub ident: Ident,
+    pub data: ast::Data<(), SystemdUnitSectionField>,
+    #[darling(rename = "section")]
+    pub section_name: String,
+}
+
+#[derive(Debug, FromField)]
+#[darling(attributes(systemd))]
+struct SystemdUnitSectionField {
+    ident: Option<Ident>,
+    entry: String,
+}
+
+impl ToTokens for SystemdUnitSection {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let SystemdUnitSection {
+            ident,
+            data,
+            section_name,
+        } = self;
+
+        let entries = data
+            .as_ref()
+            .take_struct()
+            // supports(struct_named) ensures our input will always be a struct
+            .expect("Will never be anything but a struct")
+            .fields;
+
+        // Defensively remove any brackets from the name, since we do that for the user
+        let section_name = section_name.replace(['[', ']'], "");
+
+        tokens.extend(quote! {
+            impl std::fmt::Display for #ident {
+                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                    write!(f, "[{}]\n", #section_name)?;
+
+                    #(#entries)*
+                    Ok(())
+                }
+            }
+
+        });
+    }
+}
+
+impl ToTokens for SystemdUnitSectionField {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let struct_field_name = self.ident.as_ref().expect("Should always have a name");
```

Comment:
Is there an attribute that ensures this property?

Reply:
Yes, but indirectly :)

The attribute `supports(struct_named)` on the containing struct `SystemdUnitSection` ensures the macro only supports structs with named fields.  This means it does not support tuple or newtype structs.

`SystemdUnitSectionField` has the derive attribute `FromField`, meaning it can only be "deserialized" from a struct field, and given the above `struct_named` attribute is guaranteed to have an ident/name.

Reply:
Regardless - if the developer attempted to do "the wrong thing" here, it would be a compile-time issue printing out the `expect` message.